### DISCUSSION TITLE
How to parse HTML in markdown as a single node instead of separate nodes for opening tag, content, and closing tag?

### DISCUSSION THREAD

**Discussion by rr-codes** [remarkjs/remark#745] [General] [answered]
When parsing `<sub>hi</sub>` for example with the following configuration:

```ts
import unified from 'unified';
import markdown from 'remark-parse';
import type {Block} from '@notionhq/client/build/src/api-types';
import {parseRoot} from './internal';
import gfm from 'remark-gfm';

export function parseBody(body: string): Block[] {
  const tokens = unified().use(markdown).use(gfm).parse(body);
  return parseRoot(tokens);
}
```

the generated MD AST is
```json
[
  {
    "type": "paragraph",
    "children": [
      {
        "type": "html",
        "value": "<sub>",
      },
      {
        "type": "text",
        "value": "hi",
      },
      {
        "type": "html",
        "value": "</sub>",
      }
    ]
  }
]
```

For my purposes, it would be more convenient if instead, the HTML tag and content were processed as a single node with children, like:

```json
[
  {
    "type": "paragraph",
    "children": [
      {
        "type": "html",
        "value": "sub",
        "children": [
          {
            "type": "text",
            "value": "hi",
          },
        ]
      },
    ]
  }
]
```

Is this at all possible to do? Specifically, I just want to remove all HTML tags as well as their content when parsing, which would be very easy to do if it was a single node with children.

**Comment by wooorm** [comment, marked as answer]
> I just want to remove all HTML tags as well as their content 

What about:

```markdown
<main>

All
the
markdown
content

</main>
```

---

Short answer: no. But it sounds like an [XY problem](https://xyproblem.info)

**Reply by rr-codes** [reply]
> > I just want to remove all HTML tags as well as their content
> 
> What about:
> 
> ```gfm
> <main>
> 
> All
> the
> markdown
> content
> 
> </main>
> ```

Yes, so in this case, I would want all of it to be removed.

> Short answer: no. But it sounds like an XY problem

Is there a way I could write a plugin to do so or something? Or is it just entirely impossible?

My actual goal is to convert a GFM markdown file to a Notion page body. Notion pages support some Markdown formatting but not HTML. So, the specification I have it to just create the Notion page with all HTML tags and their content removed/

**Reply by wooorm** [reply]
The way you want to achieve your goal — what you are asking — is entirely impossible. Markdown does not care whether HTML is valid or not. Take:

```
<script>

Whatever

</style>
```

With blank lines, the two tags will be seen as “HTML”. Remove those blank lines and everything will be seen as “HTML”.
There is no notion of being “inside” (or not) of HTML.

Furthermore, the grammar of HTML (and markdown too) can be expressed as:

```bnf
.*
```

([ref](https://github.com/micromark/micromark#grammar))

...meaning: anything you throw at them is HTML/markdown.

---

Depending on your notion goals, it might be better to parse HTML in markdown: go from remark, through remark-rehype, with rehype-raw, to get an HTML AST. That HTML AST then can get transformed to Notion, treating `> a` and `<blockquote><p>a<p></blockquote>` the same.

**Reply by rr-codes** [reply]
@wooorm Ah ok, thank you so much! I really appreciate it :)

**Reply by wooorm** [reply]
See also:

- https://github.com/phuctm97/ntast
- https://github.com/alexwilson/bucklespring

**Reply by rr-codes** [reply]
> Depending on your notion goals, it might be better to parse HTML in markdown: go from remark, through remark-rehype, with rehype-raw, to get an HTML AST. That HTML AST then can get transformed to Notion, treating > a and <blockquote><p>a<p></blockquote> the same.

@wooorm do you mean like this?

```ts
import unified from 'unified';
import markdown from 'remark-parse';
import type {Block} from '@notionhq/client/build/src/api-types';
import {parseRoot} from './internal';
import raw from 'rehype-raw';
import remark2rehype from 'remark-rehype';

export function parseBody(body: string): Block[] {
  const tokens = unified()
    .use(markdown)
    .use(remark2rehype, {allowDangerousHtml: true})
    .use(raw)
    .parse(body);
  return parseRoot(tokens);
}
```

?

**Reply by wooorm** [reply]
Yeah, and then transform the ast in internal!

**Reply by rr-codes** [reply]
@wooorm Would this change how `<sub>hi</sub>` for example is parsed? It doesn't appear to parse any different than how it did before. I'm a bit confused about what exactly `remark2rehype` / `rehype-raw` do differently that can help me accomplish my goal.

**Reply by wooorm** [reply]
You’re missing out on the `run` (`runSync`) step: https://github.com/unifiedjs/unified#description

```js
const processor = unified().use(...)

await processor.run(processor.parse(doc))
```
